### PR TITLE
2018 migration

### PR DESCRIPTION
**Pull request by Xaeroxe**
This is only a partial migration, this will compile for 2015 edition and 2018 edition.  The rest of the migration should probably happen in a few months.

### CODE REVIEW

**Comment by magnonellie** [comment]
Please don't merge before nalgebra 😅

**Comment by omni-viral** [comment]
Add this to see what's left.
```
#![warn(rust_2018_compatibility)]
#![warn(rust_2018_idioms)]
```

**Comment by Rhuagh** [comment]
Fix conflicts on this and we can merge.

**Comment by Xaeroxe** [comment]
Rebased, ready to merge?

**Comment by Xaeroxe** [comment]
Ok I've resolved the rustfmt conflict.  I think we can merge this now.

File: amethyst_renderer/src/pipe/pipe.rs [hunk 1]
```diff
@@ -233,11 +233,12 @@ where
 
         targets.insert("".into(), out.clone());
 
+        #[rustfmt::skip] // try is a reserved keyword in Rust 2018, must preserve keyword escape.
         let stages = self
             .stages
             .into_list()
             .fmap(BuildStage::new(fac, &targets, multisampling))
-            .try()?;
+            .r#try()?;
```
Comment (jojolepro) [inline amethyst_renderer/src/pipe/pipe.rs]: Will this stay there?

File: amethyst_renderer/src/pipe/stage.rs [hunk 2]
```diff
@@ -248,11 +248,12 @@ impl<Q> StageBuilder<Q> {
             .cloned()
             .ok_or_else(|| Error::NoSuchTarget(self.target_name.clone()))?;
 
+        #[rustfmt::skip] // try is a reserved keyword in Rust 2018, must preserve keyword escape.
         let passes = self
             .passes
             .into_list()
             .fmap(CompilePass::new(fac, &out, multisampling))
-            .try()?;
+            .r#try()?;
```
Comment (jojolepro) [inline amethyst_renderer/src/pipe/stage.rs]: and this one

File: amethyst_renderer/src/pipe/pipe.rs (hunk 1 above)
Comment (Xaeroxe) [reply amethyst_renderer/src/pipe/pipe.rs]: At least until rustfmt stops trying to kill it.

Comment (Xaeroxe) [reply amethyst_renderer/src/pipe/pipe.rs]: try is a reserved keyword in Rust 2018, so yes that will have to stay until this function's name changes.

File: amethyst_renderer/src/pass/skybox/interleaved.rs [hunk 3]
```diff
@@ -16,6 +14,8 @@ use {
     },
     set_vertex_args, ActiveCamera, Camera, Encoder, Factory, Mesh, PosTex, Shape, VertexFormat,
 };
+use gfx::pso::buffer::ElemStride;
+use glsl_layout::{mat4, Uniform};
```
Comment (magnonellie) [inline amethyst_renderer/src/pass/skybox/interleaved.rs]: Why did you move these?

File: amethyst_renderer/src/pipe/pipe.rs [hunk 4]
```diff
@@ -233,11 +233,12 @@ where
 
         targets.insert("".into(), out.clone());
 
+        #[rustfmt::skip] // try is a reserved keyword in Rust 2018, must preserve keyword escape.
```
Comment (magnonellie) [inline amethyst_renderer/src/pipe/pipe.rs]: You should probably mark this as `TODO` or `FIXME` so that it's easier to find in the future :mag:

File: examples/appendix_a/main.rs [hunk 5]
```diff
@@ -20,9 +20,9 @@ use amethyst::{
     ui::UiBundle,
     utils::application_root_dir,
 };
-use audio::Music;
-use bundle::PongBundle;
-use config::PongConfig;
+use crate::audio::Music;
+use crate::bundle::PongBundle;
+use crate::config::PongConfig;
```
Comment (magnonellie) [inline examples/appendix_a/main.rs]: Shouldn't all of these be grouped imports? :thinking:

File: rustfmt.toml [hunk 6]
```diff
@@ -1,2 +1 @@
 reorder_imports = true
-wrap_comments = false
```
Comment (magnonellie) [inline rustfmt.toml]: THAAANK YOU! :heart:

File: src/app.rs [hunk 7]
```diff
@@ -2,15 +2,15 @@
 
 use std::{error::Error as StdError, marker::PhantomData, path::Path, sync::Arc, time::Duration};
 
+use crate::shred::Resource;
```
Comment (magnonellie) [inline src/app.rs]: Why does shred in particular need to `crate::` prefix and not the other imports?

File: amethyst_renderer/src/pipe/pipe.rs (hunk 4 above)
Comment (Xaeroxe) [reply amethyst_renderer/src/pipe/pipe.rs]: Yeah I was going to do that as soon as I got more info from the rustfmt project on the issue.

File: src/app.rs (hunk 7 above)
Comment (Xaeroxe) [reply src/app.rs]: Yeah that's kind of a funny little creation.  It's a result of the fact that our shred is pulled in here: https://github.com/amethyst/amethyst/blob/72021b6c0b39a0dee13755c1547a93845a398f2c/src/lib.rs#L92

File: examples/appendix_a/main.rs (hunk 5 above)
Comment (Xaeroxe) [reply examples/appendix_a/main.rs]: yeah I suppose haha, both of the automated tools failed to catch that but we can!

File: amethyst_renderer/src/pass/skybox/interleaved.rs (hunk 3 above)
Comment (Xaeroxe) [reply amethyst_renderer/src/pass/skybox/interleaved.rs]: `cargo fmt` did.  I guess it likes them better there.

File: examples/appendix_a/main.rs (hunk 5 above)
Comment (magnonellie) [reply examples/appendix_a/main.rs]: There are more instances of this than just here btw!

File: src/app.rs (hunk 7 above)
Comment (magnonellie) [reply src/app.rs]: Ah I see.

File: amethyst_renderer/src/pass/skybox/interleaved.rs (hunk 3 above)
Comment (magnonellie) [reply amethyst_renderer/src/pass/skybox/interleaved.rs]: Does it do it even if you add a newline between the imports?

File: examples/appendix_a/main.rs (hunk 5 above)
Comment (Xaeroxe) [reply examples/appendix_a/main.rs]: I think I got them all.

File: amethyst_gltf/src/lib.rs [hunk 8]
```diff
@@ -24,8 +24,8 @@ extern crate thread_profiler;
 
 use std::{collections::HashMap, ops::Range};
 
-pub use format::GltfSceneFormat;
-use {
+pub use crate::format::GltfSceneFormat;
```
[the remark this answers is not in the document]
Comment (Xaeroxe) [reply amethyst_gltf/src/lib.rs]: I can't as that would change the publicly facing interface.  one is `pub use` and the other is `use`.

**Comment by Xaeroxe** [comment]
I added the lints for rust 2018, both compatibility and idioms.  I had to make so many changes to function signatures and types that I'm certain this PR is way more likely to generate merge conflicts now.  So as soon as this passes on CI I'd like it if we could prioritize getting this merged.

**Comment by Xaeroxe** [comment]
Huh.  `-D warnings` is going to hurt during this migration period as I can't produce code that doesn't make a warning in either Rust 2015 or Rust 2018.

**Comment by Xaeroxe** [comment]
The only way I can get this to pass on CI is if I only apply `-D warnings` to the stable compiler.  Is that an OK compromise for everyone?

**Comment by Xaeroxe** [comment]
A lot of changes have been made so I'm gonna go ahead and dismiss all reviews.  Please review again.

File: amethyst_animation/src/lib.rs [hunk 9]
```diff
@@ -45,8 +45,9 @@
 //! [ex_ani]: https://github.com/amethyst/amethyst/tree/master/examples/animation
 //! [ex_gltf]: https://github.com/amethyst/amethyst/tree/master/examples/gltf
 
-#![warn(missing_docs)]
-#![cfg_attr(feature = "cargo-clippy", allow(type_complexity))] // complex project
+#![warn(missing_docs, rust_2018_idioms, rust_2018_compatibility)]
+// I have to comment this out as right now no version of this doesn't generate warnings on both stable and beta.
+// #![allow(clippy::type_complexity)] // complex project
```
Comment (magnonellie) [inline amethyst_animation/src/lib.rs]: I would say we should remove this completely for now and properly integrate clippy in another PR.

File: amethyst_animation/src/lib.rs [hunk 10]
```diff
@@ -65,9 +66,6 @@ extern crate num_traits;
 #[macro_use]
 extern crate serde;
 
-#[cfg(feature = "profiler")]
-extern crate thread_profiler;
```
Comment (magnonellie) [inline amethyst_animation/src/lib.rs]: I probably missed a conversation somewhere, but why did you remove this?

Comment (Xaeroxe) [reply amethyst_animation/src/lib.rs]: No you didn't miss anything, I'm just trying to make the compiler happy and it said to get rid of it because it wasn't used.  That being said I'm currently battling with CI and my own compiler on how exactly thread_profiler should be brought in.

**Comment by Xaeroxe** [comment]
Since this modifies travis.yml I'm gonna

bors try

File: amethyst_animation/src/lib.rs (hunk 9 above)
Comment (Xaeroxe) [reply amethyst_animation/src/lib.rs]: Leaving the comment there is valuable as it serves as a reminder to uncomment it later imo.

**Comment by Xaeroxe** [comment]
@magnonellie r?

File: amethyst_animation/src/lib.rs (hunk 9 above)
Comment (magnonellie) [reply amethyst_animation/src/lib.rs]: I think that stuff should go in an issue and not litter the code personally.

File: amethyst_animation/src/lib.rs (hunk 10 above)
Comment (magnonellie) [reply amethyst_animation/src/lib.rs]: So are we gonna remove the profiling or what are you planning?

File: amethyst_animation/src/lib.rs (hunk 9 above)
Comment (Xaeroxe) [reply amethyst_animation/src/lib.rs]: Ok fair I'll remove it.

File: amethyst_animation/src/lib.rs (hunk 10 above)
Comment (Xaeroxe) [reply amethyst_animation/src/lib.rs]: The profiling simply wasn't in use for these crates.  I'll add it back in another PR but I considered that out of scope for this one.

File: amethyst_audio/src/lib.rs [hunk 11]
```diff
@@ -1,5 +1,6 @@
-#![warn(missing_docs)]
-#![cfg_attr(feature = "cargo-clippy", allow(type_complexity))] // complex project
+#![warn(missing_docs, rust_2018_idioms, rust_2018_compatibility)]
+// I have to comment this out as right now no version of this doesn't generate warnings on both stable and beta.
+// #![allow(clippy::type_complexity)] // complex project
```
Comment (jojolepro) [inline amethyst_audio/src/lib.rs]: re: Ellie's comment.

This one too

File: amethyst_config/src/lib.rs [hunk 12]
```diff
@@ -3,16 +3,15 @@
 
 #![crate_name = "amethyst_config"]
 #![doc(html_logo_url = "https://www.amethyst.rs/assets/amethyst.svg")]
-#![warn(missing_docs)]
+#![warn(missing_docs, rust_2018_idioms, rust_2018_compatibility)]
+// I have to comment this out as right now no version of this doesn't generate warnings on both stable and beta.
+// #![allow(clippy::type_complexity)] // complex project
```
Comment (jojolepro) [inline amethyst_config/src/lib.rs]: This one too

**Comment by Xaeroxe** [comment]
Those will be errors in Rust 2018.  I don't like it much either but I did exactly what I had to to make it compile and no more.

**Comment by torkleyy** [comment]
Any update here?

**Comment by Moxinilian** [comment]
What's wrong with the `dyn`?

**Comment by jojolepro** [comment]
Extra syntax ^^
Can't be avoided I guess.
I pinged @magnonellie on discord since xaeroxe requested a review from her.

We can merge after that.

**Comment by Xaeroxe** [comment]
no upstream conflicts, however upstream added more 2015 style code 😅 more fixes incoming, and then I'll merge.
